sycl/test-e2e/Graph/RecordReplay/transitive_queue.cpp

```diff
@@ -0,0 +1,140 @@
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+// Extra run to check for leaks in Level Zero using UR_L0_LEAKS_DEBUG
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=0 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+// Extra run to check for immediate-command-list in Level Zero
+// RUN: %if level_zero %{env SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS=1 %{l0_leak_check} %{run} %t.out 2>&1 | FileCheck %s --implicit-check-not=LEAK %}
+
+// Checks that the transitive queue recording feature is working as expected.
+// i.e. submitting a command group function to a queue that has a dependency
+// from a graph, should change the state of the queue to recording mode.
+
+#include "../graph_common.hpp"
+
+int main() {
+  queue Q1;
+  queue Q2;
+  queue Q3;
+
+  using T = int;
+
+  std::vector<T> DataA(Size), DataB(Size), DataC(Size);
+
+  std::iota(DataA.begin(), DataA.end(), 1);
+  std::iota(DataB.begin(), DataB.end(), 10);
+  std::iota(DataC.begin(), DataC.end(), 1000);
+
+  std::vector<T> ReferenceA(DataA), ReferenceB(DataB), ReferenceC(DataC);
+
+  T *PtrA = malloc_device<T>(Size, Q1);
+  T *PtrB = malloc_device<T>(Size, Q1);
+  T *PtrC = malloc_device<T>(Size, Q1);
+
+  Q1.copy(DataA.data(), PtrA, Size);
+  Q1.copy(DataB.data(), PtrB, Size);
+  Q1.copy(DataC.data(), PtrC, Size);
+  Q1.wait_and_throw();
+
+  exp_ext::command_graph Graph{Q1.get_context(), Q1.get_device()};
+
+  Graph.begin_recording(Q1);
+
+  auto GraphEventA = Q1.submit([&](handler &CGH) {
+    CGH.parallel_for(range<1>(Size), [=](item<1> Id) { PtrA[Id]++; });
+  });
+
+  // Since there is a dependency on GraphEventA which is part of a graph,
+  // this will change Q2 to the recording state.
+  auto GraphEventB = Q2.submit([&](handler &CGH) {
+    CGH.depends_on(GraphEventA);
```

Comment:
Could we have a test that uses the queue `parallel_for` shortcuts too

Reply:
Added a unit-test for the queue shortcuts